### ISSUE TITLE
PROPOSAL: `helm generate` and built-in generator/template/expansion support

### ISSUE THREAD

**Issue by technosophos** [helm/helm#309]
**tl;dr** This proposal suggests adding a `helm generate` tool analogous to the `go generate` tool, with the express purpose of providing generic support for tools that modify a manifest.

There has been a lot of talk about how we can make it easier to do certain generative steps during a various Helm phases. The ecosystem for this is growing, too. I've been trying to figure out a way to extend the plugin system to allow for this. And I'm leaning toward a method based on `go generate`.

## Use cases

As a user, I would like to...

- Generate secrets automatically at fetch or install time
- Modify environment variables (Pods, DSes, RCs, etc) at fetch or install time
- Do full templating client-side
- Do full templating server-side
- Perform simple substitutions
- Generate a manifest from some external source
- Run zero or more different tools against a manifest between fetch and install time
- Track changes to manifests locally (and VCS them)

As a Chart author, I would like to...

- Require that a user have an external tool/plugin before fetch/install
- _Allow_ a user to use a tool/plugin, but fall back on default manifests otherwise
- Create my own tool and use it in charts

## Extending the Plugin System to Allow Generating

The plugin system that exists in Helm today follows the model of Git: If `helm` receives a command that it does not know (`helm foo`), it will search on `$PATH` for a tool matching a particular nameing convention (`helm-foo`). If it finds one, it will set a number of environment variables, and then execute that command, passing it all of the remaining command line options.

This proposal suggests that we can follow a pattern like Go's `go generate` tool and use some of the features of the plugin system.

### The `helm generate` command

The command `helm generate CHART_NAME` will run a series of operations on the given chart in the workspace:

- Open each file in the `manifests/` directory
- Scan for any number of headers of the form `#helm:generate COMMAND [OPTIONS]`
- For each such located header, it will...
   - set up a number of environment variables (like Helm plugins)
   - perform shell expansion on OPTIONS, making OPTIONS'
   - and then execute the supplied COMMAND OPTIONS' inside of a shell.
   - trap and report any errors

## Modifications to `helm fetch` and `helm install`

The `helm fetch` and `helm install` commands will gain a new flag: `--generate, -g`, which will send the chart through the generation system as follows:

- `helm fetch` will copy the chart from the cache to the workspace, and then run the generator
- `helm install` will copy the chart from the cache _if necessary_, run the generator, and (if no errors occurred during generation) run `kubectl create`.

## Open Questions

Q: Should manifest files modify themselves in place? (e.g. should a `helm generate` overwrite the file in which the `helm:generate` directive exists?)

Q: Should generate iterate on all subdirectories inside of a chart, rather than just over `manifests/`? (The advantage being that chart authors could create other directories to house templated files.

I am strongly inclined toward a 'yes' answer on this one.

Q: What default tools should we include, if any?

There are already some good secrets management tools that might be good to include. Does that make sense? Are there others? Should they just be "examples"?

Q: What path(s) should be searched for COMMAND as found in the `helm:generate` line? Is `$PATH` sufficient?

**Comment by technosophos** [comment]
At one time or another I've talked to @arschles @gabrtv @jstrachan @jackgr @sgoings @mattfarina @adamreese @michelleN @jchauncey @mboersma @gerred about various aspects of templating. Any thoughts or input?

And to quickly answer a question I might not have made clear above, this system would definitely allow for the local generator to connect to remote network resources like OpenShift or DM.

Hearing the k8s team suggest how they might do things with kubectl, I think this method would work fine with that one, too. It would even be possible to run `kubectl create --dry-run -f ...` to run things through that system and then store the output locally (if someone wanted such behavior).

**Comment by arschles** [comment]
@technosophos I have a use case right now for simple variable substitution in a k8s secret and a quite a few k8s replication controllers. we also just discussed during standup the prospect of piping a value for a secret through a `base64` function a-la Go templates.

anyway, perhaps simple var substitution would be suitable for now? using the SSL certs example, such a manifest could look like the following:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: some-ssl-cert
type: Opaque
data:
  access-cert: ${ACCESS_CERT}
  access-pem: ${ACCESS_PEM}
```

... and `${ACCESS_CERT}` and `${ACCESS_PEM}` could be drawn from env vars, command line flags, or a file (in some well defined order). note that there is no "logic" (like base64 encoding, etc...) here, just substitution.

it feels like this solves a few use cases for us, and lets us start down the templating path, but we don't have to commit to full blown generators or template logic just yet

**Comment by alibazaar** [comment]
@technosophos What's the status of this feature? We would love to use this feature in production and it will be so much useful. If it's not ready yet is there anyway I could help the development and get it ready?

**Comment by technosophos** [comment]
@alibazaar It's now committed on head. Here are the docs: https://github.com/helm/helm/blob/master/docs/generate-and-template.md
